src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs

```diff
@@ -0,0 +1,329 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language.Syntax
+{
+    internal static class MarkupElementRewriter
+    {
+        public static RazorSyntaxTree AddMarkupElement(RazorSyntaxTree syntaxTree)
+        {
+            var rewriter = new AddMarkupElementRewriter();
+            var rewrittenRoot = rewriter.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        public static RazorSyntaxTree RemoveMarkupElement(RazorSyntaxTree syntaxTree)
+        {
+            var rewriter = new RemoveMarkupElementRewriter();
+            var rewrittenRoot = rewriter.Visit(syntaxTree.Root);
+
+            var newSyntaxTree = RazorSyntaxTree.Create(rewrittenRoot, syntaxTree.Source, syntaxTree.Diagnostics, syntaxTree.Options);
+            return newSyntaxTree;
+        }
+
+        private class AddMarkupElementRewriter : SyntaxRewriter
+        {
+            // From http://dev.w3.org/html5/spec/Overview.html#elements-0
+            private static readonly HashSet<string> VoidElements = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
+            {
+                "area",
+                "base",
+                "br",
+                "col",
+                "command",
+                "embed",
+                "hr",
+                "img",
+                "input",
+                "keygen",
+                "link",
+                "meta",
+                "param",
+                "source",
+                "track",
+                "wbr"
+            };
+
+            private readonly Stack<TagBlockTracker> _startTagTracker = new Stack<TagBlockTracker>();
+
+            private TagBlockTracker CurrentTracker => _startTagTracker.Count > 0 ? _startTagTracker.Peek() : null;
+
+            private string CurrentStartTagName => CurrentTracker?.TagName;
+
+            public override SyntaxNode Visit(SyntaxNode node)
+            {
+                node = base.Visit(node);
+                
+                if (node != null)
+                {
+                    node = RewriteNode(node);
+                }
+
+                return node;
+            }
+
+            private SyntaxNode RewriteNode(SyntaxNode node)
+            {
+                if (node.IsToken)
+                {
+                    // Tokens don't have children.
+                    return node;
+                }
+
+                _startTagTracker.Clear();
```

Comment:
In practice is this ever not already clear?

Reply:
It should be. Just being extra defensive.